#UTS



*   Install library sklearn



In [420]:
!pip install sklearn

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/




*   Import library pandas



In [421]:
import pandas as pd



*   Load file dataset dari github menggungakan fungsi dari library pandas



In [422]:
df = pd.read_csv("https://raw.githubusercontent.com/ThoriqFathu/prosaindata/main/dataset-baru.csv")
df

Unnamed: 0                                            Abstrak   Kategori
0             0  Sistem  informasi  akademik  (SIAKAD) merupaka...        RPL
1             1  Berjalannya koneksi jaringan komputer dengan l...        RPL
2             2  Web server adalah sebuah perangkat lunak serve...        RPL
3             3  Penjadwalan  kuliah  di  Perguruan  Tinggi  me...  KOMPUTASI
4             4  Seiring perkembangan teknologi yang ada diduni...        RPL
..          ...                                                ...        ...
800         804  Investasi saham selama ini memiliki resiko ker...  KOMPUTASI
801         805  Information Retrieval (IR) merupakan pengambil...  KOMPUTASI
802         806  Klasifikasi citra merupakan proses pengelompok...  KOMPUTASI
803         807  Identifikasi atribut pejalan kaki merupakan sa...  KOMPUTASI
804         808  Topik deteksi objek telah menarik perhatian ya...  KOMPUTASI

[805 rows x 3 columns]



*   Memisahkan kolom target (kategori) dengan kolom abstrak yang nantinya akan dijadikan feature/attribut



In [423]:
text = df['Abstrak'].values.tolist()
Y = df['Kategori'].values



*   Melakukan ekstraksi fitur menggunakan fungsi TF-IDF dari library sklearn
*   Hasil ekstraksi fitur dijadikan dataframe menggunakan fungsi dari library pandas





In [424]:
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf_vectorizer = TfidfVectorizer()
tfidf_separate = tfidf_vectorizer.fit_transform(text)

df_tfidf = pd.DataFrame(
    tfidf_separate.toarray(), columns=tfidf_vectorizer.get_feature_names_out(), index=df.index
)
X = df_tfidf.values
df_tfidf

00  000  00001  0001  000505341  000506436  000792353  00086  00089  \
0    0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   
1    0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   
2    0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   
3    0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   
4    0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   
..   ...  ...    ...   ...        ...        ...        ...    ...    ...   
800  0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   
801  0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   
802  0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   
803  0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   
804  0.0  0.0    0.0   0.0        0.0        0.0        0.0    0.0    0.0   

     001  ...   zf  zona  zone  zoning  zoom  zucara  zungu   zâ   âµ   â¼  
0    0.0  ...  0.0   0.0   0.0     0.0   0.0     0.0    0.0  0.0  0.0  0.0  
1    0.0  ...  0.0   0.0   0.0     0.0   0.0     0.0    0.0  0.0  0.0  0.0  
2    0.0  ...  0.0   0.0   0.0     0.0   0.0     0.0    0.0  0.0  0.0  0.0  
3    0.0  ...  0.0   0.0   0.0     0.0   0.0     0.0    0.0  0.0  0.0  0.0  
4    0.0  ...  0.0   0.0   0.0     0.0   0.0     0.0    0.0  0.0  0.0  0.0  
..   ...  ...  ...   ...   ...     ...   ...     ...    ...  ...  ...  ...  
800  0.0  ...  0.0   0.0   0.0     0.0   0.0     0.0    0.0  0.0  0.0  0.0  
801  0.0  ...  0.0   0.0   0.0     0.0   0.0     0.0    0.0  0.0  0.0  0.0  
802  0.0  ...  0.0   0.0   0.0     0.0   0.0     0.0    0.0  0.0  0.0  0.0  
803  0.0  ...  0.0   0.0   0.0     0.0   0.0     0.0    0.0  0.0  0.0  0.0  
804  0.0  ...  0.0   0.0   0.0     0.0   0.0     0.0    0.0  0.0  0.0  0.0  

[805 rows x 9167 columns]



*   Melakukan spilt dataset menjadi data training sebesar 80% dan data test sebesar 20% menggunkan library sklearn



In [425]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split( 
    X, Y, test_size = 0.2, random_state = 100)
print("Jumlah Data training : ", len(X_train))
print("Jumlah Data test : ", len(X_test))

Jumlah Data training :  644
Jumlah Data test :  161




**Melakukan skenario uji coba dengan mereduksi fitur dengan n component sebanyak 2,4,6,8,10**

*   n component = 2





In [426]:
from sklearn.decomposition import PCA

pca = PCA(n_components =2)

X_train_2 = pca.fit_transform(X_train)
X_test_2 = pca.transform(X_test)

pca.explained_variance_ratio_

array([0.01938057, 0.0132981 ])

In [427]:
from sklearn.neighbors import KNeighborsClassifier
classifier = KNeighborsClassifier(n_neighbors=5)
classifier.fit(X_train_2, y_train)

KNeighborsClassifier()

In [428]:
from sklearn.metrics import accuracy_score
y_pred = classifier.predict(X_test_2)
print (" KNN Accuracy : ",
    accuracy_score(y_test,y_pred)*100)

 KNN Accuracy :  72.67080745341616




*   n component = 4



In [429]:
pca = PCA(n_components =4)

X_train_4 = pca.fit_transform(X_train)
X_test_4 = pca.transform(X_test)

pca.explained_variance_ratio_

array([0.01938057, 0.01329801, 0.01235305, 0.00985703])

In [430]:
classifier = KNeighborsClassifier(n_neighbors=5)
classifier.fit(X_train_4, y_train)

KNeighborsClassifier()

In [431]:
y_pred = classifier.predict(X_test_4)
print (" KNN Accuracy : ",
    accuracy_score(y_test,y_pred)*100)

 KNN Accuracy :  75.77639751552795




*   n component = 6



In [432]:
pca = PCA(n_components = 6)
X_train_6 = pca.fit_transform(X_train)
X_test_6 = pca.transform(X_test)

pca.explained_variance_ratio_

array([0.01938057, 0.01329809, 0.01235289, 0.00985917, 0.00907997,
       0.00809378])

In [433]:
classifier = KNeighborsClassifier(n_neighbors=5)
classifier.fit(X_train_6, y_train)

KNeighborsClassifier()

In [434]:
y_pred = classifier.predict(X_test_6)
print (" KNN Accuracy : ",
    accuracy_score(y_test,y_pred)*100)

 KNN Accuracy :  79.5031055900621




*   n component = 8



In [435]:
pca = PCA(n_components = 8)

X_train_8 = pca.fit_transform(X_train)
X_test_8 = pca.transform(X_test)

pca.explained_variance_ratio_

array([0.01938057, 0.01329812, 0.01235308, 0.00985906, 0.0090823 ,
       0.00812318, 0.00768888, 0.00709752])

In [436]:
classifier = KNeighborsClassifier(n_neighbors=5)
classifier.fit(X_train_8, y_train)

KNeighborsClassifier()

In [437]:
y_pred = classifier.predict(X_test_8)
print (" KNN Accuracy : ",
    accuracy_score(y_test,y_pred)*100)

 KNN Accuracy :  82.6086956521739




*   n component = 10



In [438]:
pca = PCA(n_components = 10)

X_train_10 = pca.fit_transform(X_train)
X_test_10 = pca.transform(X_test)

pca.explained_variance_ratio_

array([0.01938057, 0.01329812, 0.0123531 , 0.00985935, 0.00908283,
       0.00812288, 0.00769172, 0.00709651, 0.0067795 , 0.00654995])

In [439]:
classifier = KNeighborsClassifier(n_neighbors=5)
classifier.fit(X_train_10, y_train)

KNeighborsClassifier()

In [440]:
y_pred = classifier.predict(X_test_10)
print (" KNN Accuracy : ",
    accuracy_score(y_test,y_pred)*100)

 KNN Accuracy :  80.74534161490683




*   Tanpa reduksi



In [441]:
classifier = KNeighborsClassifier(n_neighbors=5)
classifier.fit(X_train, y_train)

KNeighborsClassifier()

In [442]:
y_pred = classifier.predict(X_test)
print (" KNN Accuracy : ",
    accuracy_score(y_test,y_pred)*100)

 KNN Accuracy :  83.22981366459628


**KESIMPULAN**

*   Berdasarkan hasil skenario uji coba dapat disimpulkan bahwasanya dengan melakukan reduksi dimensi n componen sebanyak 8 dapat menghasilkan akurasi yang tidak jauh berbeda dengan n componen sebanyak 9167 dengan nilai akurasi diatas 80% menggunakan model KNN dengan nilai k sebesar 5
*   Dengan demikian metode PCA dapat diterapkan untuk mempercepat proses komputasi



